src/i_pcsound.c

```diff
@@ -0,0 +1,448 @@
+        return;
+    }
+
+    alSource3i(callback_source, AL_POSITION, 0, 0, -1);
+    alSourcei(callback_source, AL_SOURCE_RELATIVE, AL_TRUE);
+    alSourcei(callback_source, AL_ROLLOFF_FACTOR, 0);
+    alSourcePlay(callback_source);
+    if (alGetError() != AL_NO_ERROR)
+    {
+        I_Printf(VB_ERROR, "RegisterCallback: Failed start playback.");
+        return;
+    }
+}
+
+static void UnregisterCallback(void)
+{
+    alSourceStop(callback_source);
+    alDeleteSources(1, &callback_source);
+    alDeleteBuffers(1, &callback_buffer);
+}
+
+static boolean I_PCS_ReinitSound(void)
+{
+    mixing_freq = SND_SAMPLERATE;
+
+    current_freq = 0;
+    current_remaining = 0;
+
+    if (!alBufferCallbackSOFT)
+    {
+        RegisterCallback(BufferCallback);
+        I_AtExitPrio(UnregisterCallback, true, "UnregisterCallback",
+                     exit_priority_first);
+    }
+
+    if (!sound_lock)
+    {
+        sound_lock = SDL_CreateMutex();
+    }
+
+    return true;
+}
+
+static boolean I_PCS_InitSound(void)
+{
+    if (!I_OAL_InitSound())
+    {
+        return false;
+    }
+
+    I_PCS_ReinitSound();
+
+    return true;
+}
+
+
+static boolean I_PCS_AdjustSoundParams(const mobj_t *listener, const mobj_t *source,
+                                      int chanvol, int *vol, int *sep, int *pri)
+{
+    alSourcef(callback_source, AL_GAIN, (float)snd_SfxVolume / 15);
+    return true;
+}
+
+static void I_PCS_UpdateSoundParams(int channel, int volume, int separation)
+{
+    // no-op
+}
+
+static boolean I_PCS_StartSound(int channel, sfxinfo_t *sfx, int pitch)
+{
+    boolean result;
+
+    if (IsDisabledSound(sfx))
+    {
+        return true;
+    }
+
+    if (SDL_LockMutex(sound_lock) < 0)
+    {
+        return false;
+    }
+
+    result = CachePCSLump(sfx);
+
+    if (result)
+    {
+        current_sound_handle = channel;
+    }
+
+    SDL_UnlockMutex(sound_lock);
+
+    if (result)
+    {
+        return true;
+    }
+    else
+    {
+        return false;
+    }
+}
+
+static void I_PCS_StopSound(int channel)
+{
+    if (SDL_LockMutex(sound_lock) < 0)
+    {
+        return;
+    }
+
+    // If this is the channel currently playing, immediately end it.
+
+    if (current_sound_handle == channel)
+    {
+        current_sound_remaining = 0;
+    }
+
+    SDL_UnlockMutex(sound_lock);
+}
+
+static boolean I_PCS_SoundIsPlaying(int channel)
+{
+    if (channel != current_sound_handle)
+    {
+        return false;
+    }
+
+    return current_sound_lump != NULL && current_sound_remaining > 0;
+}
+
+const sound_module_t sound_pcs_module =
+{
+    I_PCS_InitSound,
+    I_PCS_ReinitSound,
+    I_OAL_AllowReinitSound,
+    I_OAL_UpdateUserSoundSettings,
+    I_OAL_CacheSound,
+    I_PCS_AdjustSoundParams,
+    I_PCS_UpdateSoundParams,
+    NULL,
+    I_PCS_StartSound,
+    I_PCS_StopSound,
+    I_PCS_SoundIsPlaying,
+    I_OAL_ShutdownSound,
+    I_OAL_DeferUpdates,
+    I_OAL_ProcessUpdates,
```

Comment:
@ceski-1 I'm thinking of removing `I_DeferUpdates` and `I_ProcessUpdates` from the `sound_module_t` interface and making them `static` in `i_oalsound.c`. I also want to redo `I_ShutdownSound`. What do you think?

Reply:
The idea of defer/process updates is that we can control exactly when the listener/sources update. Normally, OpenAL Soft updates a sound parameter as soon as possible which, in theory, can lead to a partial update that sounds strange. There is a good explanation [here](https://openal-soft.org/openal-extensions/SOFT_deferred_updates.txt). Maybe it matters more for slow computers.

For Woof, I originally considered defer/process at a lower level but then I saw that `S_UpdateSounds()` loops through each sound channel, so it would be pointless. Also, `S_UpdateSounds()` relies on variables and functions at the `s_sound.c` and `i_sound.c` level so it's not easy to move it entirely into `i_oalsound.c`.

Reply:
Right, making `I_DeferUpdates` and `I_ProcessUpdates` `static` will not work, I was wrong.

The idea is to reduce the `sound_module_t` interface and call `I_OAL_*` functions directly in `i_sound.c`. Is it worth doing such a rearrangement?

Reply:
So for example `sound_module->InitSound()` would become `I_OAL_InitSound()` in `i_sound.c`? Yeah, it makes sense to directly call `I_OAL_*` for functions that the sound modules will never have a need to modify. The extra layer was there to "isolate" OAL but if that's not important, I don't see a problem with simplifying.